examples/DataApi-NFT/template/modules/shared_data/src/main/scala/com/my/nft_example/shared_data/validations/Validations.scala

```diff
@@ -1,98 +1,128 @@
+          val validateNFTDescription = validateStringMaxSize(update.description, 64, "description")
+          val validateNFTMetadata = validateMapMaxSize(update.metadata, 15, "metadata")
+
+          validateNFTUrlValid
+            .productR(validateNFTName)
+            .productR(validateNFTDescription)
+            .productR(validateNFTMetadata)
         }
     }
-  }
 
-  def transferCollectionValidations(update: TransferCollection, maybeState: Option[DataState[NFTUpdatesState, NFTUpdatesCalculatedState]]): IO[DataApplicationValidationErrorOr[Unit]] = {
+  def transferCollectionValidations[F[_] : Async](
+    update    : TransferCollection,
+    maybeState: Option[DataState[NFTUpdatesState, NFTUpdatesCalculatedState]]
+  ): F[DataApplicationValidationErrorOr[Unit]] =
     maybeState match {
+      case None => valid.pure[F]
       case Some(state) =>
-        val validateProvidedCollection = validateIfProvidedCollectionExists(update, state)
-        val validateFromAddress = validateIfFromAddressIsTheCollectionOwner(update, state)
+        Async[F].delay {
+          val validateProvidedCollection = validateIfProvidedCollectionExists(update, state)
+          val validateFromAddress = validateIfFromAddressIsTheCollectionOwner(update, state)
 
-        IO {
           validateProvidedCollection.productR(validateFromAddress)
         }
-      case None => IO(().validNec)
     }
 
-  }
 
-  def transferNFTValidations(update: TransferNFT, maybeState: Option[DataState[NFTUpdatesState, NFTUpdatesCalculatedState]]): IO[DataApplicationValidationErrorOr[Unit]] = {
+  def transferNFTValidations[F[_] : Async](
+    update    : TransferNFT,
+    maybeState: Option[DataState[NFTUpdatesState, NFTUpdatesCalculatedState]]
+  ): F[DataApplicationValidationErrorOr[Unit]] =
     maybeState match {
+      case None => valid.pure[F]
       case Some(state) =>
-        val validateProvidedNFT = validateIfProvidedNFTExists(update, state)
-        val validateFromAddress = validateIfFromAddressIsTheNFTOwner(update, state)
+        Async[F].delay {
+          val validateProvidedNFT = validateIfProvidedNFTExists(update, state)
+          val validateFromAddress = validateIfFromAddressIsTheNFTOwner(update, state)
 
-        IO {
           validateProvidedNFT.productR(validateFromAddress)
         }
-      case None => IO(().validNec)
     }
-  }
 
-  def mintNFTValidationsWithSignature(update: MintNFT, addresses: List[Address], state: DataState[NFTUpdatesState, NFTUpdatesCalculatedState]): IO[DataApplicationValidationErrorOr[Unit]] = {
-    val validateAddress = validateProvidedAddress(addresses, update.owner).pure[IO]
+  def mintNFTValidationsWithSignature[F[_] : Async](
+    update   : MintNFT,
+    addresses: List[Address],
+    state    : DataState[NFTUpdatesState, NFTUpdatesCalculatedState]
+  ): F[DataApplicationValidationErrorOr[Unit]] = {
+    val validateAddress = Async[F].delay(validateProvidedAddress(addresses, update.owner))
```

Comment:
In case the difference is not obvious, here's a sample test. If you uncomment the 2nd assertion you'll see it fail and see the difference.
```
test("productR") { _ =>
  import cats.syntax.all._
  import org.tessellation.currency.dataApplication.{DataApplicationValidationError, Noop}
  import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr

  case object Hoop extends DataApplicationValidationError {
    val message = "hoop"
  }

  val hoop : DataApplicationValidationErrorOr[Unit] = Hoop.invalidNec[Unit]
  val hoopF = hoop.pure[IO]

  val noop: DataApplicationValidationErrorOr[Unit] = Noop.invalidNec[Unit]
  val noopF = noop.pure[IO]

  for {
    prodF: DataApplicationValidationErrorOr[Unit] <- hoopF.productR(noopF)
    prodV: DataApplicationValidationErrorOr[Unit] = hoop.productR(noop)
    prodM: DataApplicationValidationErrorOr[Unit] <- hoopF.map2(noopF)(_.productR(_))
  } yield expect.all(
    prodM == prodV
//    , prodF == prodV
  )
}
```

Reply:
Thanks, updated